Login: Jetpack setup error handling

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #8075 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds error handling for the native Jetpack setup flow. There are 3 cases handled here:

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
When the user taps the "Log in with another account" link on the Jetpack connection web page, they can log in with a different account other than the logged-in one. I handled this edge case by:
- Updated `SiteCredentialLoginViewModel` to return the site XMLRPC URL when the login completes.
- Updated the `LoginJetpackSetupViewModel` to return the Jetpack-connected email when the setup completes.
- Updated the coordinator to navigate to WP.com account verification if the Jetpack-connected email is different from the logged-in email.

**2. Error handling for setup steps.**
- Updated `LoginJetpackSetupViewModel` with information about the setup failure, error messages and codes, and visibility for error state views.
- `checkJetpackConnection` is handled with an auto-retry mechanism that allows a maximum of 2 retries. This is to ensure that Jetpack is synced properly for the site and returns the correct connection status.
- Updated `LoginJetpackSetupView` to show appropriate contents for error states.

**3. Handling interrupted Jetpack connection.**
When the user dismisses the web view for the Jetpack connection, we should show a modal indicating that the connection step has been interrupted. A new SwiftUI view has been added for this: `LoginJetpackSetupInterruptedView`.

Due to the size of the PR, unit tests will be added in a separate PR.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Prerequisite: make sure that you have a test site without Jetpack.

**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for your store.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving Jetpack connection.
- Tap the "Log in with another account" on the web page and proceed to log in with a WP.com other than your logged-in account.
- Tap on the Authorize button on the web page, notice that the web view is then dismissed, and the Jetpack setup screens should then show that all steps are completed.
- Tap on Go to Store, notice that you are navigated to a new screen for verifying the password for the account you just entered on the web page. 
- After the login succeeds, notice that you are navigated to the home screen.

**2. Error handling for an account without permission.**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for an account **whose role is neither shop manager or administrator.**
- On the Jetpack setup screen, notice that the error state is then be displayed with a message regarding a permission issue. The retry CTA should not be available for this error. The Cancel and Get support CTAs should work properly.

**3. Error handling for issues other than permission**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- When the Jetpack setup screen appears, turn off the internet connection or intercept the response with a 5xx error. Notice that the error state will then be displayed with appropriate messages. The retry CTA should be available and work properly.

**4. Error handling for interrupted Jetpack connection**
- Log out of the app or skip login onboarding if needed.
- On the prologue screen, select "Log In" or "Continue with WordPress.com" based on the A/B test variant you get.
- Log in with your WordPress.com account.
- On the store picker, select Add a Store > Connect an existing site.
- Enter the address of your test store and tap Continue.
- On the Jetpack error screen, select Install Jetpack or Connect Jetpack.
- When the site credential login screen is presented, enter the correct credentials for a shop manager or administrator account.
- On the Jetpack setup screen, when the setup reaches the connection step, notice that a web view is presented for approving the Jetpack connection.
- Tap Cancel to dismiss the web view, notice that a new view is presented regarding the interrupted Jetpack connection.
- Tap Continue Connection, notice that the error screen is dismissed, and the web view is displayed again.
- Repeat tapping Cancel, this time tap Help, notice that the support screen is displayed.
- Tap the Cancel Installation button, notice that the installation flow is dismissed.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
**1. Jetpack connection authorized with a WP.com account other than the logged-in one.**

https://user-images.githubusercontent.com/5533851/203531637-52ce0fc7-63da-429e-ba7b-3989bd5ca038.mp4

**2. Error handling for an account without permission.**

https://user-images.githubusercontent.com/5533851/203531829-3b708d4a-9c1f-416a-8010-da30d76f55e8.mp4


**3. Error handling for issues other than permission**

https://user-images.githubusercontent.com/5533851/203531940-2e512b8a-e6be-41a8-8786-e3980c134a94.mp4

**4. Error handling for interrupted Jetpack connection**

https://user-images.githubusercontent.com/5533851/203532077-50ccd77e-5707-4b31-a90e-efb19e00cd96.mp4



---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Handle edge case where the user authorizes Jetpack with another WP.com account

itsmeichigo committed:
Fix test failure due to completion closures

selanthiraiyan committed:
JetpackSetupRequiredViewModel - Add analytics events on view did load.

selanthiraiyan committed:
JetpackSetupRequiredViewModel - unit tests for testing events on view did load.

selanthiraiyan committed:
Add analytics events for Site credentials screen.

selanthiraiyan committed:
Add tracks events for Install/Setup Jetpack screen.

selanthiraiyan committed:
LoginJetpackSetupHostingController - Add tracks events for view did load and view dismiss.

selanthiraiyan committed:
LoginJetpackSetupHostingController - Add unit tests for tracks events.

selanthiraiyan committed:
SiteCredentialLoginViewModel - Add tracks events and unit tests.

selanthiraiyan committed:
Remove unnecessary import.

selanthiraiyan committed:
SiteCredentialLoginHostingViewController - Add tracks events for view did load and dismissal.

selanthiraiyan committed:
Inject analytics for tracking and unit testing.

selanthiraiyan committed:
Track "Go to Store" button tap.

selanthiraiyan committed:
Add unit tests for tracking event of "Go to Store" button tap.

selanthiraiyan committed:
Track events when Jetpack install is success/failure.

selanthiraiyan committed:
Unit test to track events when Jetpack install is success/failure.

selanthiraiyan committed:
Add tracks events when Jetpack activation is success/failure.

selanthiraiyan committed:
Unit tests for tracks events when Jetpack activation is success/failure.

selanthiraiyan committed:
Add  tracks events when fetching Jetpack connection URL is success/failure.

selanthiraiyan committed:
Unit tests for tracks events when fetching Jetpack connection URL is success/failure.

selanthiraiyan committed:
Add tracks events when checking Jetpack connection is success/failure.

selanthiraiyan committed:
Unit tests for tracks events when checking Jetpack connection is success/failure.

selanthiraiyan committed:
Rename Jetpack setup screen related events for clarity.

itsmeichigo committed:
Add retry mechanism to LoginJetpackSetupViewModel

itsmeichigo committed:
Update JetpackInstallHeaderView with error state

itsmeichigo committed:
Update title for LoginJetpackSetupView when an error occurs

selanthiraiyan committed:
Introduce Tracks events for no matched site alert.

selanthiraiyan committed:
Add tracks events for no matched site alert.

itsmeichigo committed:
Show errors on LoginJetpackSetupView

selanthiraiyan committed:
Inject analytics for unit testing.

selanthiraiyan committed:
Add extra properties for site discovery event.

selanthiraiyan committed:
Add unit testing for site discovery event.

selanthiraiyan committed:
Inject analytics for tracking and unit testing.

selanthiraiyan committed:
Fix completion block param error

selanthiraiyan committed:
Remove whitespace.

itsmeichigo committed:
Add buttons for error state

selanthiraiyan committed:
Fix indentation.

itsmeichigo committed:
Remove auto retry for jetpack installation, activation & fetching connection url

itsmeichigo committed:
Add new image for jetpack setup interrupted

itsmeichigo committed:
Add jetpack setup interrupted view

itsmeichigo committed:
Handle connection interruption

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupInterruptedView.swift:
```diff
@@ -0,0 +1,98 @@
+    var body: some View {
+        GeometryReader { geometry in
+            ScrollView {
+                VStack {
```
Nit: Is it possible to use `ScrollableVStack` here?

itsmeichigo replied:
Good idea 👍  Updated in 8dfed739182f9daf71e1ed0e33efe09449820ac9.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -3,14 +3,27 @@ import SwiftUI
+            guard let self else { return }
+            // dismiss any presented view if possible
+            self.presentedViewController?.dismiss(animated: true)
+            authentication.presentSupport(from: self, screen: .jetpackRequired)
```
Super nit: Consider moving this to a private method just like other handlers.

itsmeichigo replied:
Updated in 1be991101591e3f49bf29444e654ded92a100db0.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -100,7 +124,7 @@ struct LoginJetpackSetupView: View {
                     Spacer()
                 }
-                .renderedIf(viewModel.currentSetupStep == nil)
+                .renderedIf(viewModel.shouldShowInitialLoadingIndicator)
```
Nice 👍

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -176,6 +271,8 @@ private extension LoginJetpackSetupView {
+        static let errorContentSpacing: CGFloat = 16
     }
 }
 
```
Super nit: Is it possible to add a view to `LoginJetpackSetupView_Previews` to show the error scenario?

itsmeichigo replied:
The errors are get-only, so I'm not sure how to mock that 😅 Please let me know if you have any suggestion.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -153,21 +210,59 @@ struct LoginJetpackSetupView: View {
+            LoginJetpackSetupInterruptedView(onSupport: supportHandler, onContinue: {
+                viewModel.jetpackConnectionInterrupted = false
+                // delay for 300ms for the dismissal of the interrupted screen to complete.
+                DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
```
Should we move `0.3` to a `Constants` enum?

itsmeichigo replied:
Yup, updated in f357e38413da7fa30d298e488831904610bf2428.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -139,8 +163,41 @@ struct LoginJetpackSetupView: View {
+                    .buttonStyle(.plain)
+                    .foregroundColor(Color(uiColor: .withColorStudio(.blue, shade: .shade50)))
 
+                }
```
Nit: Moving this to a private view will help with readability. Or kindly consider adding comments to label the view.

itsmeichigo replied:
Added comments in 20e00e190ce5448e54fac0ba650cae9c2ccd6cd4.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupView.swift:
```diff
@@ -153,21 +210,59 @@ struct LoginJetpackSetupView: View {
-            .renderedIf(viewModel.currentSetupStep == .done)
+            .renderedIf(viewModel.shouldShowGoToStoreButton)
+
+            VStack(spacing: Constants.contentVerticalSpacing) {
+                Button {
+                    // TODO: add tracks
+                    viewModel.retryAllSteps()
+                } label: {
+                    Text(viewModel.tryAgainButtonTitle)
+                }
+                .buttonStyle(PrimaryButtonStyle())
+                .renderedIf(viewModel.hasEncounteredPermissionError == false)
+
+                Button {
+                    // TODO: add tracks
+                    cancellationHandler()
+                } label: {
+                    Text(Localization.cancelInstallation)
+                }
+                .buttonStyle(SecondaryButtonStyle())
+            }
```
Nit: Same as above, moving this to a private view will help with readability. Or kindly consider adding comments.

itsmeichigo replied:
Also updated in 20e00e190ce5448e54fac0ba650cae9c2ccd6cd4.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/LoginJetpackSetupViewModel.swift:
```diff
@@ -8,15 +9,64 @@ final class LoginJetpackSetupViewModel: ObservableObject {
+
+    /// Whether the setup failed. This will be observed by `LoginJetpackSetupView` to present error modal.
+    @Published private(set) var setupFailed: Bool = false
+    @Published private(set) var setupErrorMessage: String = ""
+    @Published private(set) var setupErrorSuggestion: String = ""
+    @Published private(set) var errorCode: Int?
```
Nit: If all of this all related, would it make sense to combine these into a `Struct`?

itsmeichigo replied:
Good point! Updated in f6037222df9e2a65eaf670a1b89720f195608946.

selanthiraiyan reviewed (approved):
Great work, Huong. 🎉 

All the scenarios described in the PR work as expected. ✅ 

I left a few non-blocking nits and questions.

selanthiraiyan committed:
Track error only after retrying fails.

selanthiraiyan committed:
New event for tracking when user authorises Jetpack using different WPCOM account.

selanthiraiyan committed:
Track when user authorises Jetpack using different WPCOM account.

selanthiraiyan committed:
Track the actual error along with tracks event.

selanthiraiyan committed:
Track the actual error along with tracks event.

selanthiraiyan committed:
Track the actual error along with tracks event.

selanthiraiyan committed:
Track the actual error along with tracks event.

selanthiraiyan committed:
Track the actual error along with tracks event.

itsmeichigo committed:
Use ScrollableVStack for LoginJetpackSetupInterruptedView

itsmeichigo committed:
Extract presentSupport to a separate method

itsmeichigo committed:
Move delay time for interrupted connection action handlers

itsmeichigo committed:
Add comments for LoginJetpackSetupView components

itsmeichigo committed:
Add new subtype for error detail

itsmeichigo committed:
Add dark mode assets

itsmeichigo commented:
Thank you @selanthiraiyan for the review! I updated the code following your suggestions above.

I also added assets for dark mode in ab1613f2b6fd7e28327422b4d5ee6cbeead85038 - sorry for cramping this here since CI might take a long time to handle several PRs 😅 .

Screenshots for dark mode:

| Jetpack setup | Jetpack connection | Interrupted connection |
| ----- | ----- | ----- |
| <img src="https://user-images.githubusercontent.com/5533851/203715218-2eba62ce-8f95-45e3-a7f7-9f2ee97df414.png" width=320 /> | <img src="https://user-images.githubusercontent.com/5533851/203715260-898033e5-2db6-4a35-b7c0-f247200ff077.png" width=320 /> | <img src="https://user-images.githubusercontent.com/5533851/203715285-060218e0-58f4-462b-8451-0e4e26228dbc.png" width=320 /> |

Updated light mode asset for interrupted connection:

<img src="https://user-images.githubusercontent.com/5533851/203715552-1847af70-f12d-4c00-ba2f-b8ef299d6420.png" width=320 />

selanthiraiyan commented:
Thanks for the updates. This still LGTM. 🚢

itsmeichigo committed:
Fix text color of error state view

itsmeichigo committed:
Handle retry for no WPCom user issue

selanthiraiyan committed:
Merge pull request #8204 from woocommerce/issue/8075-tracks-event-for-different-wpcom-account

Login - Jetpack setup - Tracks event for authorization using different WPCOM account

selanthiraiyan committed:
Merge branch 'issue/8075-jetpack-setup-error-handling' into issue/8075-add-tracks

selanthiraiyan committed:
Fix error during solving merge conflict.

selanthiraiyan committed:
Merge pull request #8187 from woocommerce/issue/8075-add-tracks

Login - Jetpack Setup: Add tracks events
